src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -11612,7 +11612,10 @@ public bool Join(in LocalState other)
 
                 for (int i = 1; i < Capacity; i++)
                 {
-                    var oldValue = GetValue(i);
+                    // An unreachable state may contain uninitialized `(false, false)` values,
+                    // but it can be turned into a reachable state during a join (see above).
+                    // During that process we can't call GetValue on it.
+                    var oldValue = oldReachable ? GetValue(i) : NullableFlowState.NotNull;
```

Comment:
It occurs to me that the entire Join/Meet loops could be made a little "smarter" around reachability:

For Join:
1. if 'other' is unreachable, do nothing.
2. if 'this' is unreachable, blind-copy 'other' into 'this'.
3. otherwise do the ordinary Join loop.

For Meet:
1. if 'other' is unreachable, blind-copy 'other' into 'this'.
2. if 'this' is unreachable, do nothing.
3. otherwise do the ordinary Meet loop.

However, it's not required/urgent to do this, and if it doesn't feel quite right in terms of complexity/assumptions/... feel free to skip it and just keep the current logic.

Reply:
I agree we could do that. What would be the motivation for the change in your mind: readability or performance?
Merging/joining with one unreachable state doesn't seem that common (see examples I had to write to hit this), so I'm not sure this optimization is worth the extra logic in Merge/Join (note we also need to compute the `hasChanged` bit to return). 
So I'd rather punt on that change for now as I think it will make this code more complicated, not simpler.